Circuit breaker plugin

Camillarhi opened the pull request:
Add Circuit Breaker sidecar integration for LND nodes

- Integrated Circuit Breaker as sidecar container alongside LND nodes
- Added configuration options (enabled flag, configurable httpPort)
- Created user documentation with setup examples and web UI access instructions
- Implemented shared volume mounting for TLS cert and macaroon access

Camillarhi committed:
Add circuit breaker plugin

pinheadmz commented:
weird error trying to `warnet deploy resources/networks/hello/` but might be related to https://github.com/bitcoin-dev-project/warnet/pull/689 like, `plugin.py` needs `chmod a+x` or something.

```
Queuing postNetwork plugin command: hello with {'entrypoint': '../../plugins/hello', 'helloTo': 'postNetwork!', 'podName': 'hello-post-network'}
Starting postNetwork plugins
Completed postNetwork plugins
Queuing postDeploy plugin command: hello with {'entrypoint': '../../plugins/hello', 'podName': 'hello-post-deploy', 'helloTo': 'postDeploy!'}
Queuing postDeploy plugin command: simln with {'entrypoint': '../../plugins/simln', 'activity': '[{"source": "tank-0003-ln", "destination": "tank-0005-ln", "interval_secs": 1, "amount_msat": 2000}]'}
Queuing postDeploy plugin command: circuitbreaker with {'entrypoint': '../../plugins/circuitbreaker', 'podName': 'circuitbreaker-pod', 'rpcserver': '172.29.34.166:10009', 'httplisten': '0.0.0.0:9235'}
Starting postDeploy plugins
Process Process-11:
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
    ~~~~~~~~^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthewzipkin/Desktop/work/warnet/src/warnet/process.py", line 7, in run_command
    raise Exception(result.stderr)
Exception: bash: line 1: resources/networks/hello/../../plugins/circuitbreaker/plugin.py: Permission denied

```

Camillarhi commented:
> weird error trying to `warnet deploy resources/networks/hello/` but might be related to #689 like, `plugin.py` needs `chmod a+x` or something.
> 
> ```
> Queuing postNetwork plugin command: hello with {'entrypoint': '../../plugins/hello', 'helloTo': 'postNetwork!', 'podName': 'hello-post-network'}
> Starting postNetwork plugins
> Completed postNetwork plugins
> Queuing postDeploy plugin command: hello with {'entrypoint': '../../plugins/hello', 'podName': 'hello-post-deploy', 'helloTo': 'postDeploy!'}
> Queuing postDeploy plugin command: simln with {'entrypoint': '../../plugins/simln', 'activity': '[{"source": "tank-0003-ln", "destination": "tank-0005-ln", "interval_secs": 1, "amount_msat": 2000}]'}
> Queuing postDeploy plugin command: circuitbreaker with {'entrypoint': '../../plugins/circuitbreaker', 'podName': 'circuitbreaker-pod', 'rpcserver': '172.29.34.166:10009', 'httplisten': '0.0.0.0:9235'}
> Starting postDeploy plugins
> Process Process-11:
> Traceback (most recent call last):
>   File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/multiprocessing/process.py", line 313, in _bootstrap
>     self.run()
>     ~~~~~~~~^^
>   File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/multiprocessing/process.py", line 108, in run
>     self._target(*self._args, **self._kwargs)
>     ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/matthewzipkin/Desktop/work/warnet/src/warnet/process.py", line 7, in run_command
>     raise Exception(result.stderr)
> Exception: bash: line 1: resources/networks/hello/../../plugins/circuitbreaker/plugin.py: Permission denied
> ```

Yeah, I believe this issue is related to https://github.com/bitcoin-dev-project/warnet/pull/689. To resolve it, I `ran chmod a+x` on the relevant plugin files and converted line Endings with `dos2unix`. That got everything working for me.

willcl-ark commented:
<!--e57a25ab6845829454e8d69fc972939a-->

The following sections might be updated with supplementary metadata relevant to reviewers and maintainers.

<!--174a7506f384e20aa4161008e828411d-->
### Conflicts
No conflicts as of last run.

macgyver13 commented on resources/charts/bitcoincore/charts/lnd/templates/pod.yaml:
```diff
@@ -56,12 +73,41 @@ spec:
           name: config
           subPath: tls.cert
-    {{- if .Values.circuitBreaker }}
     - name: circuitbreaker
```
I would recommend making circuitbreaker optional using extraContainers added in [#691](https://github.com/bitcoin-dev-project/warnet/pull/691)

Camillarhi replied:
Alright, thanks. This has been done

macgyver13 commented on resources/plugins/circuitbreaker/plugin.py:
```diff
@@ -0,0 +1,204 @@
+#         subprocess.run(["kubectl", "cp", f"{node}:/root/.lnd/tls.cert", "./tls.cert"], check=True)
+#         subprocess.run(["kubectl", "cp", f"{node}:/root/.lnd/data/chain/bitcoin/regtest/admin.macaroon", "./admin.macaroon"], check=True)
+#         subprocess.run(["kubectl", "create", "secret", "generic", f"lnd-tls-cert-{node_name}", "--from-file=tls.cert=./tls.cert"], check=True)
+#         subprocess.run(["kubectl", "create", "secret", "generic", f"lnd-macaroon-{node_name}", "--from-file=admin.macaroon=./admin.macaroon"], check=True)
```
remove this commented block

Camillarhi replied:
Thanks but this implementation has been removed

macgyver13 commented on resources/charts/bitcoincore/charts/lnd/templates/pod.yaml:
```diff
@@ -19,6 +19,27 @@ spec:
   {{- end }}
   securityContext:
     {{- toYaml .Values.podSecurityContext | nindent 4 }}
+  initContainers:
```
If initContainers is ultimately required it should be included on a conditional basis

Camillarhi replied:
I added this `enableInitContainers` to the values. If the initContainers logic is not required (e.g., if the macaroon is already in binary format or if the application does not need it), the Pod spec remains lightweight.

macgyver13 commented on resources/charts/bitcoincore/charts/lnd/templates/pod.yaml:
```diff
@@ -44,6 +65,8 @@ spec:
         {{- toYaml .Values.resources | nindent 8 }}
       volumeMounts:
+        - mountPath: /shared-data
+          name: shared-data
```
use values.volumeMounts to include this information if required

macgyver13 commented on resources/charts/bitcoincore/charts/lnd/templates/pod.yaml:
```diff
@@ -56,12 +79,12 @@ spec:
+    {{- end }}
   volumes:
+    - name: shared-data
+      emptyDir: {}
```
use values.volumes to include this information if required

macgyver13 commented on resources/charts/bitcoincore/charts/lnd/values.yaml:
```diff
@@ -132,3 +132,12 @@ config: ""
+# This volume is mounted at the specified mountPath and can be used for inter-container communication.
+sharedVolume:
+  name: shared-data
+  mountPath: /shared-data
```
These new values should be defined elsewhere like the network.yaml or node-defaults.yaml files

Camillarhi replied:
I added `enableInitContainers` and `sharedVolume` to the `network.yaml` file. These settings are now configured per node, so they only apply to nodes that explicitly define them, like `tank-0003`. Nodes without these settings will skip them entirely, avoiding errors and keeping things flexible.

pinheadmz commented on resources/charts/bitcoincore/charts/lnd/templates/pod.yaml:
```diff
@@ -58,6 +58,9 @@ spec:
           subPath: tls.cert
+        - mountPath: /root/.lnd/macaroon.hex
+          name: config
+          subPath: MACAROON_HEX
```
This shouldn't be necessary because lnd will deterministically generate the expected macaroon file shortly after startup:

https://github.com/bitcoin-dev-project/warnet/blob/8e7b1359cb9f831f744707c74668f1c69ab007d8/resources/charts/bitcoincore/charts/lnd/values.yaml#L93-L95

Camillarhi replied:
Alright I will look into this and make the changes

Camillarhi replied:
@pinheadmz I have updated the macaroon file to be retrieved from a shared volume between the two containers(lnd and circuitbreaker

pinheadmz commented on resources/networks/hello/network.yaml:
```diff
@@ -32,7 +32,24 @@ nodes:
+            - "--network=regtest"
+            - "--rpcserver=localhost:10009"
+            - "--tlscertpath=/tls.cert"
+            - "--macaroonpath=/macaroon.hex"
```
as long as circuitbreaker either waits for lnd to start, or otherwise keeps trying to start until lnd has generated its macaroon file, you can expect it to be at ` /root/.lnd/data/chain/bitcoin/regtest/admin.macaroon` (assuming regtest)

Camillarhi replied:
I will use a dynamic network in the path

Camillarhi replied:
Hello @pinheadmz  the network path is now being pulled as a global helm variable

pinheadmz commented on resources/networks/hello/network.yaml:
```diff
@@ -32,7 +32,24 @@ nodes:
+          imagePullPolicy: IfNotPresent
+          args:
+            - "--network=regtest"
+            - "--rpcserver=localhost:10009"
```
network and port may change depending on the user config, you should pull these as global helm variables similar to how lnd pulls it from the bitcoin pod config.

Camillarhi replied:
Alright, I will update the values to be pulled from the global helm variable

Camillarhi replied:
Hello @pinheadmz  the network path is now being pulled as a global helm variable

pinheadmz commented on resources/networks/hello/network.yaml:
```diff
@@ -32,7 +32,24 @@ nodes:
-
+      extraContainers:
+        - name: circuitbreaker
+          image: camillarhi/circuitbreaker:latest
```
I'm getting an `ImagePullBackOff` error. Did you push your image to dockerhub? Is there some reason why you don't want to use https://hub.docker.com/r/carlakirkcohen/circuitbreaker

Camillarhi replied:
I built the image using ubuntu on windows, I think that is why you could not pull it. I used mine because of the change I made on the circuit breaker code to accept a hex value for the `macaroon` file.

Camillarhi replied:
I have updated the image to use this(https://hub.docker.com/r/carlakirkcohen/circuitbreaker) now.

pinheadmz reviewed (commented):
I noticed a few things to check. Coudn't get the branch to run locally probably just  because of the image error?

Also checkout the legacy branch for how we did integration tests for circuitbreaker:

https://github.com/bitcoin-dev-project/warnet/blob/legacy/test/ln_test.py#L34-L39

pinheadmz commented on resources/charts/bitcoincore/charts/lnd/templates/pod.yaml:
```diff
@@ -58,16 +58,37 @@ spec:
         - mountPath: /root/.lnd/tls.cert
           name: config
           subPath: tls.cert
+        - name: shared-volume
+          mountPath: /root/.lnd/
```
I think you might not actually need this shared volume at all. Because lnd and circuitbreaker are in the same pod, they already share a filesystem. So you'd just need to configure circuitbreaker to read from the right location. There's also apparently an issue getting the network value into the config.

When I `warnet deploy resources/networks/hello` I see this error in the circuitbreaker container:

```
2025-04-29T13:54:23.979Z    INFO    Circuit Breaker starting    {"version": ""}
2025-04-29T13:54:23.979Z    INFO    Opening database    {"path": "/root/.circuitbreaker/circuitbreaker.db"}
2025-04-29T13:54:24.007Z    INFO    Applied migrations    {"count": 4}
2025-04-29T13:54:24.007Z    ERROR    Unexpected exit    {"err": "unable to read macaroon path (check the network setting!): open /root/.lnd/data/chain/bitcoin/regtest/admin.macaroon: no such file or directory"}
main.main
    /src/main.go:152
runtime.main
    /usr/local/go/src/runtime/proc.go:250
Stream closed EOF for default/tank-0003-ln (circuitbreaker)

```

pinheadmz replied:
@Camillarhi did you try removing the shared volume?

Camillarhi replied:
I was looking into other comments. I will look into this as soon as I can

Camillarhi replied:
@pinheadmz , After testing both approaches, I've confirmed that a shared volume is necessary for reliable file access between containers in the same pod. While containers in a pod share the same network namespace, they maintain separate filesystems unless explicitly shared through volumes.

pinheadmz commented on resources/charts/bitcoincore/charts/lnd/values.yaml:
```diff
@@ -132,3 +132,8 @@ config: ""
+
+circuitbreaker:
+  enabled: false  # Default to disabled
+  image: carlakirkcohen/circuitbreaker:attackathon-ln_51_attacker
```
I led you a little astray on this sorry the tag to pull, for now, is `attackathon-test`.

Camillarhi replied:
The latest tag is `attackathon-ln_51_attacker` do I still update it to `attackathon-test`?

pinheadmz replied:
yeah the tags with numbers like 10, 51 etc have pre-loaded data for specific networks, other purposes

Camillarhi replied:
Alright thanks, I will make the change

pinheadmz reviewed (commented):
Looking good, thanks for the update. Be sure to check out the legacy ln test that covered circuitbreaker: https://github.com/bitcoin-dev-project/warnet/blob/legacy/test/ln_test.py

I don't think the `hello` network is very useful actually, and it would be better to add the circuit breaker plugin to one of the networks in `test/data` so it can actually be run and tested by ci.

Camillarhi commented:
> Looking good, thanks for the update. Be sure to check out the legacy ln test that covered circuitbreaker: https://github.com/bitcoin-dev-project/warnet/blob/legacy/test/ln_test.py
> 
> I don't think the `hello` network is very useful actually, and it would be better to add the circuit breaker plugin to one of the networks in `test/data` so it can actually be run and tested by ci.

Thanks. I have added circuitbreaker to a network and some unit tests for it has been included as well

macgyver13 commented on resources/charts/bitcoincore/charts/lnd/templates/configmap.yaml:
```diff
@@ -38,7 +38,8 @@ data:
     -----END EC PRIVATE KEY-----
-  MACAROON_HEX: 0201036c6e6402f801030a1062beabbf2a614b112128afa0c0b4fdd61201301a160a0761646472657373120472656164120577726974651a130a04696e666f120472656164120577726974651a170a08696e766f69636573120472656164120577726974651a210a086d616361726f6f6e120867656e6572617465120472656164120577726974651a160a076d657373616765120472656164120577726974651a170a086f6666636861696e120472656164120577726974651a160a076f6e6 ...
+  MACAROON_HEX: |
+    0201036c6e6402f801030a1062beabbf2a614b112128afa0c0b4fdd61201301a160a0761646472657373120472656164120577726974651a130a04696e666f120472656164120577726974651a170a08696e766f69636573120472656164120577726974651a210a086d616361726f6f6e120867656e6572617465120472656164120577726974651a160a076d657373616765120472656164120577726974651a170a086f6666636861696e120472656164120577726974651a160a076f6e636861696e120 ...
```
Nit: no change is required here, remove this change.

Camillarhi replied:
Alright, I will remove the change

macgyver13 commented on resources/charts/bitcoincore/charts/lnd/templates/pod.yaml:
```diff
@@ -58,16 +58,37 @@ spec:
     {{- with .Values.extraContainers }}
     {{- toYaml . | nindent 4 }}
     {{- end }}
+    {{- if .Values.circuitbreaker.enabled }}
+    - name: circuitbreaker
+      image: {{ .Values.circuitbreaker.image | quote }}
+      imagePullPolicy: IfNotPresent
+      args:
+        - "--network={{ .Values.global.chain }}"
+        - "--rpcserver=localhost:{{ .Values.RPCPort }}"
+        - "--tlscertpath=/tls.cert"
+        - "--macaroonpath=/root/.lnd/data/chain/bitcoin/{{ .Values.global.chain }}/admin.macaroon"
+        - "--httplisten=0.0.0.0:{{ .Values.circuitbreaker.httpPort }}"
+      volumeMounts:
+        - name: shared-volume
+          mountPath: /root/.lnd/
+        - name: config
+          mountPath: /tls.cert
+          subPath: tls.cert
+    {{- end }}
```
Was there an issue including circuitbreaker as extraContainers container?

Camillarhi replied:
Yes, I was not able to access the value of the port and network variables as the `extraContainers` is being parsed by a YAML parser before Helm gets to process the templates.

pinheadmz replied:
Hm the whole point of those `extraContainers` was for the cb plugin ... :-/

Camillarhi replied:
Yes, I understand, but I could not access a dynamic network or port while using it. The current approach also gives the option of enabling circuitbreaker if needed.

macgyver13 replied:
@pinheadmz I assume it is acceptable to define network(chain) in network.yaml or node-defaults.yaml based on user desired scenario.  if so then extraContainers should be possible...

macgyver13 commented on test/ln_basic_test.py:
```diff
@@ -120,6 +130,103 @@ def scenario_open_channels(self):
+            )
+        except subprocess.CalledProcessError as e:
+            self.log.error(f"Failed to create service: {e.stderr}")
+            raise
```
You could add service.yaml definition instead of creating service on the fly...
<img width="535" alt="image" src="https://github.com/user-attachments/assets/d770961f-50c1-4edd-ae25-2edf050d500a" />

Camillarhi replied:
Thanks. I will do this instead

pinheadmz replied:
I think this is overkill. See the legacy test, you can just execute an API request form inside the cb container. (this exact code wont work any more but maybe gives a good hint)

https://github.com/bitcoin-dev-project/warnet/blob/01195c2fa497cd95b7a9bf41cf87a8bfe8d577c2/test/ln_test.py#L34-L39

Camillarhi replied:
I looked into the legacy test but was unable to replicate it or access `circuitbreaker` through `localhost(127.0.0.1)` without port forwarding.  I’ll do more research to figure out how to properly access it, but I’m also open to suggestions.

pinheadmz replied:
you could add a util funciton to k8s.py to execute an arbitrary command inside a container (surprised we dont have that alreaady actually....) and then execute the command `curl 127.0.0.1/whatever/cb/api...`

pinheadmz replied:
oh right or do something like this in the test:

```
cmd = f"kubectl -n {namespace} exec {tank} --container {BITCOINCORE_CONTAINER} -- bitcoin-cli {method} {' '.join(map(str, params))}"
return run_command(cmd)
```

Camillarhi replied:
Thanks. I will try this

Camillarhi replied:
Hello @pinheadmz , the command has been updated as you suggested and now uses `run_command` imported from `warnet.process`.

This is ready for another review

pinheadmz commented:
I still think the test can be simplified a great deal, IMO you dont need those helper functions, observe

deploy the test network outside of the test
```

$ warnet deploy test/data/ln/
```

execute api calls inside the circuit breaker container and retrieve the output
```
$ kubectl exec tank-0003-ln -c circuitbreaker -- wget -qO - 127.0.0.1:9235/api/limits

{"limits":[],"defaultLimit":{"maxHourlyRate":"3600","maxPending":"5","mode":"MODE_FAIL"}}

$ kubectl exec tank-0003-ln -c circuitbreaker -- wget -qO - 127.0.0.1:9235/api/info

{"nodeKey":"02a88cbef03eeefa74e8ac963b64ff24813d16dd4035d91f4547ad8c40440c516f","nodeAlias":"tank-0003-ln","nodeVersion":"0.19.0-beta commit=v0.19.0-beta","version":""}

```

Camillarhi commented:
> I still think the test can be simplified a great deal, IMO you dont need those helper functions, observe
> 
> deploy the test network outside of the test
> 
> ```
> 
> $ warnet deploy test/data/ln/
> ```
> 
> execute api calls inside the circuit breaker container and retrieve the output
> 
> ```
> $ kubectl exec tank-0003-ln -c circuitbreaker -- wget -qO - 127.0.0.1:9235/api/limits
> 
> {"limits":[],"defaultLimit":{"maxHourlyRate":"3600","maxPending":"5","mode":"MODE_FAIL"}}
> 
> $ kubectl exec tank-0003-ln -c circuitbreaker -- wget -qO - 127.0.0.1:9235/api/info
> 
> {"nodeKey":"02a88cbef03eeefa74e8ac963b64ff24813d16dd4035d91f4547ad8c40440c516f","nodeAlias":"tank-0003-ln","nodeVersion":"0.19.0-beta commit=v0.19.0-beta","version":""}
> ```

Thanks for the feedback! I've simplified the test as suggested, removed the complex service setup/port forwarding and replaced it with direct kubectl exec calls to query the circuit breaker API. Let me know if you'd like any further tweaks

pinheadmz reviewed (approved):
ACK 5905f4523a348c193164abb9c85c7a37d0f871c6
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 5905f4523a348c193164abb9c85c7a37d0f871c6
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmhAUXQACgkQ5+KYS2KJ
yTo2thAAzCOMBFbrOoLzhryqKoz2RWAWIkfpt7tZLbPI48sF7cJ2ikBqIYsWvBdx
UUvArwthp8qJRpcYP5VDOCbPqXyQBH5f+pGRduRq4ilEMK4c/QYDyKdPM3VqwWer
edKSGdSIQUi5B/kHwcu+TL+gJPkfbRdv1uvLID551S1yyqPhaCBVazEDzEwg9UFz
2JsZeWmpVYYxvVVkFRh1D3yJn9l5zHBchRPWH4klVHEdvVRoXzq++CMTiv6AxPf/
6DxngBa4FW8m6yxJJyZh0JucymMgrCt2kvP1E11FjF65EXxuMhHUoe1+BPLdrdll
fKGE6id8BsvyTFTWmsZwxOCLkNeLfEAaAsvoIaf9gL01iZZ3dD7w+SPwNvlW3MzE
yufzKFl3GvWSw3pPiMQjdFkwFFs5Yg5SAlhk15nW0GhsINtDpuEDAI+eE6HHGC2V
hlh0BZvhQAWDS36cPIPgQ038fS/uqGE6D4BRRzFlXRSm+rx/ugsx+j9jIlDALy/W
+UsItHB/EqlTne7jkaqEHAZaQ3uDhHSBMchWIylm7XwWXTVTARnsV490wlH3nwM/
1don7jyhxO1za8LwkAH330yRpQPolm9w2oDzXdNoIMbQZl3oqGkV2ljoUlK4qNhi
kAlbrx5SHjhfWuAtWKqKqIyB5UB0J8DWHWjc8cfrbVcc+IF/5Eg=
=WLKL
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

pinheadmz commented:
Thank you so much! great work!!!!

Camillarhi commented:
> Thank you so much! great work!!!!

Thanks a lot for the guidance as well. It was great working on this
